Import results data:  There are 1500 rows of data

In [18]:
import numpy as np
import pandas as pd

df = pd.read_csv('results.csv')
df = df.iloc[:,1:]
df = df.rename(index = str, columns = {'conf_matrix': 'f_measure'})
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1500 entries, 0 to 1499
Data columns (total 31 columns):
QI1               1500 non-null float64
QI2               1500 non-null float64
QI3               1500 non-null float64
QI4               1500 non-null float64
QI5               1500 non-null float64
QI6               1500 non-null float64
QI7               1500 non-null float64
QI8               1500 non-null float64
QI9               1500 non-null float64
QI10              1500 non-null float64
QI11              1500 non-null float64
QI12              1500 non-null float64
QI13              1500 non-null float64
QI14              1500 non-null float64
QI15              1500 non-null float64
QI16              1500 non-null float64
QIs               1500 non-null float64
cluster_method    1500 non-null float64
clusters          1500 non-null float64
2k                1500 non-null float64
3k                1500 non-null float64
4k                1500 non-null float64
5k                

In [19]:
#Group data by grouping variables
df_grouped = df.groupby(['QIs','cluster_method','clusters'])

In [34]:
#Turn off annoying autoscroll feature
%%javascript
IPython.OutputArea.auto_scroll_threshold = 300;

<IPython.core.display.Javascript object>

In [35]:
#Make sure all rows are displayed
pd.set_option('display.max_rows', None)

#don't display last two columns
data_list = df.columns[19:-2]

#display group means
df_grouped[data_list].mean()

2k        3k        4k        5k  \
QIs  cluster_method clusters                                           
2.0  0.0            2.0       0.999750  0.999750  0.999374  0.999374   
                    10.0      0.995995  0.991239  0.987860  0.984856   
                    30.0      0.938673  0.884606  0.833166  0.786108   
                    100.0     0.824781  0.729662  0.665457  0.613392   
                    500.0     0.539049  0.406383  0.345932  0.307384   
     1.0            2.0       1.000000  1.000000  1.000000  1.000000   
                    10.0      0.998373  0.993867  0.986733  0.965707   
                    30.0      0.718523  0.436671  0.309387  0.256320   
                    100.0     0.204255  0.076095  0.034418  0.016395   
                    500.0     0.253817  0.118148  0.050563  0.016521   
3.0  0.0            2.0       0.999374  0.999374  0.999374  0.998874   
                    10.0      0.962954  0.934168  0.903379  0.878849   
                    30.0      0.734668  0.621026  0.545181  0.494118   
                    100.0     0.615645  0.466959  0.381727  0.329161   
                    500.0     0.144556  0.062954  0.037046  0.026033   
     1.0            2.0       1.000000  1.000000  1.000000  1.000000   
                    10.0      0.656696  0.375094  0.236170  0.184606   
                    30.0      0.156195  0.059074  0.030914  0.016395   
                    100.0     0.010638  0.001126  0.000000  0.000000   
                    500.0     0.004255  0.000000  0.000000  0.000000   
4.0  0.0            2.0       0.999750  0.999499  0.999124  0.996621   
                    10.0      0.902753  0.837171  0.780851  0.739299   
                    30.0      0.422153  0.285232  0.211640  0.164581   
                    100.0     0.186233  0.093116  0.067960  0.054944   
                    500.0     0.041302  0.006008  0.001502  0.000000   
     1.0            2.0       0.999750  0.999750  0.999750  0.999750   
                    10.0      0.301877  0.115645  0.057822  0.033292   
                    30.0      0.016270  0.000000  0.000000  0.000000   
                    100.0     0.005131  0.000375  0.000000  0.000000   
                    500.0     0.002003  0.000000  0.000000  0.000000   
5.0  0.0            2.0       0.999750  0.999249  0.999249  0.998248   
                    10.0      0.881727  0.795620  0.732541  0.671965   
                    30.0      0.399625  0.252691  0.180225  0.142678   
                    100.0     0.033792  0.005757  0.000501  0.000000   
                    500.0     0.004631  0.000375  0.000000  0.000000   
     1.0            2.0       0.999249  0.998999  0.996746  0.994743   
                    10.0      0.047810  0.005257  0.001502  0.000000   
                    30.0      0.000501  0.000000  0.000000  0.000000   
                    100.0     0.000000  0.000000  0.000000  0.000000   
                    500.0     0.000000  0.000000  0.000000  0.000000   
6.0  0.0            2.0       0.998874  0.997372  0.995870  0.994368   
                    10.0      0.674969  0.511765  0.413392  0.331289   
                    30.0      0.151690  0.075845  0.045432  0.029412   
                    100.0     0.022904  0.001627  0.000501  0.000000   
                    500.0     0.000501  0.000000  0.000000  0.000000   
     1.0            2.0       0.994994  0.984731  0.970088  0.952065   
                    10.0      0.016020  0.000751  0.000000  0.000000   
                    30.0      0.000000  0.000000  0.000000  0.000000   
                    100.0     0.000000  0.000000  0.000000  0.000000   
                    500.0     0.000000  0.000000  0.000000  0.000000   
7.0  0.0            2.0       0.997121  0.995870  0.994368  0.990864   
                    10.0      0.643554  0.474844  0.372716  0.313642   
                    30.0      0.169962  0.095870  0.069587  0.059074   
                    100.0     0.002753  0.000000  0.000000  0.000000   
    

**Overview**:  Final dataset as analyzed had 16 attributes.  Above are the grouped results of this analysis.  The grouping variables are QIs, cluster_method, and clusters.  
1. QIs is the number of columns selected for clustering (2-16).  
2. Cluster_method is the binning strategy (fixed-member = 0, fixed-width = 1).  
3. Clusters is the level of generalization/coarsening provided by the cluster method (2,10,30,100,500).  For instance, a value of 100 means that each value in a selected columns is assigned to 1 of 100 discrete values.

All together now.  A row in the above table with say (QI = 15, Cluster_method = 0, and clusters = 2) represents the **average** data of 10 runs with those settings.  Each run randomly selected 15 of the 16 columns to cluster, then clustered each using fixed-member (quantiled) clustering into 2 values per column. You can see this provides great anonymity (all the way up to 10-anonymity), though our accuracy of the model is prob just out of range to meet the 95%CI threshhold

Notes on results values (2k - 10k).  As stated previously each row is an average result.  A 2k value represents the average ratio of rows that had **at least** (2-1) one other row with matching QI attribute values. A 10k value on the same row is generated from the same data and represents the ratio of rows that had at least (10-1) nine other rows with matching QI attributes.  Also, note that this algorithm only checks for matches within the clustered QI columns, the unclustered columns are not included.  This is why when the number of QIs is high, anonymity is much worse as it  becomes more difficult to create matching rows. 

Notes on accuracy and f-measure.  I say we disregard f-measure.  Accuracy is the accuracy of the average random forest model trained and executed on the settings shown by the grouping variables. The model is trained on the grouped QI data **and** the remaining ungrouped columns.  It is trained on 10% of the records (randomly chosen) and evaluated on 10% of the records (randomly chosen and dissimilar). Each random forest is composed of 20 trees who vote to determine the prediction. Bottom line is we see great durability in our model.  Only with extreme clustering - once with QIs very high (13-16), cluster_method = 0 and with low clusters - do we see it's performance start to drop off below 95% accuracy.  For reference the unclustered model has accuracy of 99.8%

**Results**:

1. Show which clustering technique provides better anonymity:  Fixed-member(cluster_method = 0) binning provides better anonymity than fixed-width (cluster_method = 1) binning.  This is true across the board.  The effect of each varies with the number of QIs (more QIs = worse anonymity) and the cluster sizes (more clusters = more granular data and worse anonymity), but the overall relationship remains (fixed-member better for this data-set)

2. Show how much better fixed-width binning is:  For instance,  3 QIs with fixed-member binning can handle up to 100 clusters and still provide better than .6 2-anonymity while fixed width-binning needs to be heavily clustered (10 clusters per QI) to provide >.6 2-anonymity.

3. Show how model results (accuracy) vary with number of QIs:  Model performs worse with more QIs.  This is due to the fact that more of the columns are clustered.

4. Show how model results vary with clustering method.  Not sure if one performs better in this regard, maybe method 1 gives better accuracy for given set of QIs and clusters (you can see a slight difference at high QIs when clusters is 2).

5. Show how model results vary with cluster size.  When there are few clusters (few = large) we see model results perform worse.  This is particularly visible when there at high QIS and true for both clustering methods. 

6. Putting it all together - show which combination of grouping variables provides best trade-off between anonymity and model performance.  If we set an anonymity cut-off at .9 and accuracy cut-off of .95 (per engels), what are the grouping variables that allows us the most QIs.  Answer:  With 10 QIs we are able to meet both thresholds.  This is using method 0, and very coarse clustering (clusters = 2).  Note the anonymity threshold is actually met for all values of k (2 - 10).  What about the other clustering technique? Cluster method 1 can meet both thresholds at QIs = 8, but only for 2-anonymity.  One has to go down to QIs = 5 to see method 1 provide 10-anonymity and meet both thresholds.  
